File: src/lib/api/changeEvents.ts
import { getActiveBackendClient } from "@/lib/bridge/activeBackendClient";
import type { ChangeEvent, UpsertChangeEventInput } from "@/lib/types/changeEvents";

function changeEventsClient() {
  return getActiveBackendClient().changeEvents;
}

export function listChangeEvents(): Promise<ChangeEvent[]> {
  return changeEventsClient().listChangeEvents();
}

export function clearChangeEvents(): Promise<void> {
  return changeEventsClient().clearChangeEvents();
}

export function listChangeEventsForStation(stationId: string): Promise<ChangeEvent[]> {
  return changeEventsClient().listChangeEventsForStation(stationId);
}

export function upsertChangeEvent(input: UpsertChangeEventInput): Promise<ChangeEvent> {
  return changeEventsClient().upsertChangeEvent(input);
}

export function markChangeEventRead(id: string): Promise<ChangeEvent> {
  return changeEventsClient().markChangeEventRead(id);
}

export function markChangeEventsRead(ids: string[]): Promise<ChangeEvent[]> {
  const uniqueIds = Array.from(new Set(ids.filter(Boolean)));
  if (uniqueIds.length === 0) {
    return Promise.resolve([]);
  }

  return changeEventsClient().markChangeEventsRead(uniqueIds);
}

export function dismissChangeEvent(id: string): Promise<ChangeEvent> {
  return changeEventsClient().dismissChangeEvent(id);
}

export function resolveChangeEvent(id: string): Promise<ChangeEvent> {
  return changeEventsClient().resolveChangeEvent(id);
}

File: src-tauri/src/application/command_facades/change_events.rs
use std::sync::Arc;

use crate::{
    application::{changes::ChangeService, error::ApplicationError, pagination::PageLimit},
    models::change_events::{ChangeEvent, UpsertChangeEventInput},
};

#[derive(Clone)]
pub(crate) struct ChangeEventsCommandFacade {
    changes: Arc<ChangeService>,
}

impl ChangeEventsCommandFacade {
    pub(crate) fn new(changes: Arc<ChangeService>) -> Self {
        Self { changes }
    }

    pub(crate) async fn list_change_events(
        &self,
        station_id: Option<&str>,
        limit: PageLimit,
    ) -> Result<Vec<ChangeEvent>, ApplicationError> {
        self.changes.list(station_id, limit).await
    }

    pub(crate) async fn clear_change_events(&self) -> Result<(), ApplicationError> {
        self.changes.clear().await
    }

    pub(crate) async fn upsert_change_event(
        &self,
        input: UpsertChangeEventInput,
    ) -> Result<ChangeEvent, ApplicationError> {
        self.changes.upsert(input).await
    }

    pub(crate) async fn mark_change_event_read(
        &self,
        id: String,
    ) -> Result<ChangeEvent, ApplicationError> {
        self.changes.mark_read(id).await
    }

    pub(crate) async fn mark_change_events_read(
        &self,
        ids: Vec<String>,
    ) -> Result<Vec<ChangeEvent>, ApplicationError> {
        self.changes.mark_many_read(ids).await
    }

    pub(crate) async fn dismiss_change_event(
        &self,
        id: String,
    ) -> Result<ChangeEvent, ApplicationError> {
        self.changes.dismiss(id).await
    }

    pub(crate) async fn resolve_change_event(
        &self,
        id: String,
    ) -> Result<ChangeEvent, ApplicationError> {
        self.changes.resolve(id).await
    }
}
